### ISSUE TITLE
SX1262 Sleep

### ISSUE THREAD

**Issue by lvplummer** [jgromes/RadioLib#740]
My board is very simple - Arduino UNO chip and crystal with no regulator or serial interrace - battery powered.  I made two changes to your transmit example without interrupts:  pin assignments to match my board and serial baud rate changed to 115200.

I turned on debug and compiled your blocking SX1262 transmit example.  Here are the results from that:

18:03:36.088 -> [SX1262] Initializing ... Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
18:03:36.088 -> 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 | SX1261 V2D 2D02.
18:03:36.088 -> 
18:03:36.088 -> M	SX126x
18:03:36.088 -> Symbol length: 4719 ms
18:03:36.136 -> Symbol length: 4719 ms
18:03:36.136 -> Symbol length: 4719 ms
18:03:36.136 -> success!
18:03:36.136 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:03:36.278 -> success!
18:03:36.278 -> [SX1262] Datarate:	567.64 bps
18:03:37.273 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:03:37.462 -> success!
18:03:37.462 -> [SX1262] Datarate:	567.64 bps
18:03:38.455 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:03:38.643 -> success!
18:03:38.643 -> [SX1262] Datarate:	567.64 bps
18:03:39.633 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:03:39.822 -> success!
18:03:39.822 -> [SX1262] Datarate:	567.64 bps
18:03:40.811 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:03:40.999 -> success!

I then inserted a radio.sleep() just before the one second delay.  

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }

  radio.sleep();
  // wait for a second before transmitting again
  delay(1000);
}

Here are the results of that:

18:09:34.496 -> [SX1262] Initializing ... Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
18:09:34.496 -> 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 | SX1261 V2D 2D02.
18:09:34.496 -> 
18:09:34.496 -> M	SX126x
18:09:34.496 -> Symbol length: 4719 ms
18:09:34.544 -> Symbol length: 4719 ms
18:09:34.544 -> Symbol length: 4719 ms
18:09:34.544 -> success!
18:09:34.544 -> [SX1262] Transmitting packet ... Timeout in -8704 us
18:09:34.686 -> success!
18:09:34.686 -> [SX1262] Datarate:	567.70 bps
18:09:35.711 -> [SX1262] Transmitting packet ... Timed out waiting for GPIO pin, is it connected?
18:09:36.714 -> failed, code -705
18:09:37.708 -> [SX1262] Transmitting packet ... Timed out waiting for GPIO pin, is it connected?
18:09:38.694 -> failed, code -705
18:09:39.723 -> [SX1262] Transmitting packet ... Timed out waiting for GPIO pin, is it connected?
18:09:40.703 -> failed, code -705

The first transmission before the first sleep works correctly.   Subsequent transmission fail.

**Comment by jgromes** [comment]
Thanks for the report, I was able to find this one quickly. There was a bug in waking up the SX126x device after it was put to sleep (the Arduino just timed out waiting for the BUSY signal to go low, which is not going to happen when the device is sleeping). I added a check to make sure that it will wakeup. Tested on Uno + SX1268, thanks for reporting!

**Comment by lvplummer** [comment]
Thanks for your great software and also the quick support.

However, your solution seems to have added one second somewhere.  My
application requires long battery life which means that CPU and
radio current during sleep must be very low AND a transmit cycle needs to
be very short.

Your sleep fix has caused my transmit cycle to go from 190 milliseconds to
1.2 seconds.

**Comment by lvplummer** [comment]
Seems that the additional second is not coming from sleep but rather from
transmit which I assume does the wakeup.

**Comment by jgromes** [comment]
I see - that's caused by the timeout while waiting for BUSY, which is one second.

I added additional parameter to the `standby()` method, so if the one second waiting time is critical for you, you can wakeup the module immediately like this:

```c++
radio.sleep();
// radio sleeping here
(...)

// OK, time to wake up
radio.standby(RADIOLIB_SX126X_STANDBY_RC, true);

// radio in standby, transmit
radio.transmit(...)
```

**Comment by lvplummer** [comment]
Wow!  Your real time support is amazing.

I just tried your solution and it works.

I hope not to have to bug you for a while.

Larry

**Comment by lvplummer** [comment]
Sorry to still be a pain.

I have three simple battery powered remote telemetry gadgets that send data
one per minute.  I tried your fix on all three.  Two work correctly and one
does not.

I ran your modified example sketch with the standby fix on all three in
debug mode.  Here are the results from one that works and one that
doesn't.  I tried switching the UNO chip and the issue is only related to
the SX1262 module.  I am guessing some timing parameter is right on the edge

A reset (begin)

14:01:48.846 -> [SX1262] Initializing ... Found SX126x:
RADIOLIB_SX126X_REG_VERSION_STRING:
14:01:48.846 -> 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 |
SX1261 V2D 2D02.
14:01:48.846 ->
14:01:48.846 -> M SX126x
14:01:48.846 -> Symbol length: 4719 ms
14:01:48.846 -> Symbol length: 4719 ms
14:01:48.846 -> Symbol length: 4719 ms
14:01:48.846 -> success!
14:01:48.846 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:01:49.035 -> success!
14:01:49.035 -> [SX1262] Datarate: 567.64 bps
14:01:50.027 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:01:50.216 -> success!
14:01:50.216 -> [SX1262] Datarate: 567.64 bps
14:01:51.210 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:01:51.387 -> success!

---------------------------------------------------

14:04:03.425 -> [SX1262] Initializing ... Found SX126x:
RADIOLIB_SX126X_REG_VERSION_STRING:
14:04:03.425 -> 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 |
SX1261 V2D 2D02.
14:04:03.425 ->
14:04:03.425 -> M SX126x
14:04:03.425 -> Symbol length: 4719 ms
14:04:03.425 -> Symbol length: 4719 ms
14:04:03.425 -> Symbol length: 4719 ms
14:04:03.425 -> success!
14:04:03.425 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:04:03.661 -> timeout!
14:04:04.648 -> [SX1262] Transmitting packet ... failed, code -2
14:04:05.647 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:04:05.836 -> success!
14:04:05.836 -> [SX1262] Datarate: 550.33 bps
14:04:06.821 -> [SX1262] Transmitting packet ... failed, code -2
14:04:07.855 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:04:07.994 -> success!
14:04:07.994 -> [SX1262] Datarate: 550.29 bps
14:04:09.042 -> [SX1262] Transmitting packet ... failed, code -2
14:04:10.011 -> [SX1262] Transmitting packet ... Timeout in -8704 us
14:04:10.197 -> success!
14:04:10.197 -> [SX1262] Datarate: 550.31 bps
14:04:11.201 -> [SX1262] Transmitting packet ... failed, code -2

**Comment by jgromes** [comment]
That's interesting, couple of thoughts:

1. Is there any difference between the SX126x module that works and the others? I'm guessing  the rest of the hardware (the MCU, power etc.) is the same for all three.
2. Could you run the module that does not work with both DEBUG and VERBOSE enabled?

**Comment by lvplummer** [comment]
1.  I have a total of five transmitter boards - four in one band and one in
another.  All are identical and very simple - UNO DIP chip, crystal, two
caps, DS3231 RTC, and two AAA batteries.  The CPU is in a DIP socket so I
swapped the CPU chip and the issue aligned with the radio module.  The
modules for the two bands are from two different manufacturers.  NiceRF for
the four 915 MHz modu;es and another Chinese company that I can't remember
the name of for the one 433 MHz band.

2.  I tried turning on both DEBUG and VERBOSE where the output display was
the same that I sent before.

Even though the unit with the" bad" module works except when I use sleep, I
think it is time to call this module defective and move on.  If I see
something similar from the boards that are known good, I will let you know.

Larry

**Comment by jgromes** [comment]
OK, let me know if the issue pops up again, I'd like to get to the bottom of this.

**Comment by lvplummer** [comment]
I am up for helping anyway I can.  Let me know if you have ideas or code to
test.

**Comment by lvplummer** [comment]
With no sleep and DEBUG and VERBOSE on, here is the output of one cycle:

10:48:19.232 -> [SX1262] Initializing ... CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 3 20
10:48:19.515 -> DATR 0 A2 0 53 0 58 0 31 0 32 0 36 0 31 0 20 0 56 0 32 0 44
0 20 0 32 0 44 0 30 0 32 0 0
10:48:19.515 ->
10:48:19.515 -> Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
10:48:19.515 -> 0000320 53 58 31 32 36 31 20 56 32 44 20 32 44 30 32 00 |
SX1261 V2D 2D02.
10:48:19.515 ->
10:48:19.515 -> M SX126x
10:48:19.515 -> CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 17
10:48:19.515 -> DATR 0 A2 0 0 0 0
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 97
10:48:19.515 -> DATW 0 A2 0 A2 1 A2 40 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8F
10:48:19.515 -> DATW 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8A
10:48:19.515 -> DATW 1 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 93
10:48:19.515 -> DATW 20 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 88
10:48:19.515 -> DATW 3 A2 16 A2 A A2 0 A2 0 A2 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 2
10:48:19.515 -> DATW 43 A2 FF A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8
10:48:19.515 -> DATW 0 A2 0 A2 0 A2 0 A2 0 A2 0 A2 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 89
10:48:19.515 -> DATW 7F A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> Symbol length: 4719 ms
10:48:19.515 -> CMD 8B
10:48:19.515 -> DATW 9 A2 6 A2 3 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 7 40
10:48:19.515 -> DATW 14 A2 24 A2
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 7 36
10:48:19.515 -> DATR 0 A2 0 D
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 7 36
10:48:19.515 -> DATW D A2
10:48:19.515 ->
10:48:19.515 -> CMD 8C
10:48:19.515 -> DATW 0 A2 8 A2 0 A2 FF A2 1 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 96
10:48:19.515 -> DATW 1 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 8 E7
10:48:19.515 -> DATW 18 A2
10:48:19.515 ->
10:48:19.515 -> CMD 9D
10:48:19.515 -> DATW 1 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 7 36
10:48:19.515 -> DATR 0 A2 0 D
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 7 36
10:48:19.515 -> DATW D A2
10:48:19.515 ->
10:48:19.515 -> CMD 8C
10:48:19.515 -> DATW 0 A2 8 A2 0 A2 FF A2 1 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> Symbol length: 4719 ms
10:48:19.515 -> CMD 8B
10:48:19.515 -> DATW 9 A2 6 A2 3 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> Symbol length: 4719 ms
10:48:19.515 -> CMD 8B
10:48:19.515 -> DATW 9 A2 4 A2 3 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 98
10:48:19.515 -> DATW 6B A2 6F A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 86
10:48:19.515 -> DATW 1B A2 20 A2 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 8 D8
10:48:19.515 -> DATR 0 A2 0 DE
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 8 D8
10:48:19.515 -> DATW DE A2
10:48:19.515 ->
10:48:19.515 -> CMD 1D 8 E7
10:48:19.515 -> DATR 0 A2 0 18
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 95
10:48:19.515 -> DATW 4 A2 7 A2 0 A2 1 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8E
10:48:19.515 -> DATW A A2 4 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 8 E7
10:48:19.515 -> DATW 18 A2
10:48:19.515 ->
10:48:19.515 -> success!
10:48:19.515 -> [SX1262] Transmitting packet ... CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 80
10:48:19.515 -> DATW 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> Timeout in -8704 us
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 7 36
10:48:19.515 -> DATR 0 A2 0 D
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 7 36
10:48:19.515 -> DATW D A2
10:48:19.515 ->
10:48:19.515 -> CMD 8C
10:48:19.515 -> DATW 0 A2 8 A2 0 A2 C A2 1 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8
10:48:19.515 -> DATW 2 A2 1 A2 0 A2 1 A2 0 A2 0 A2 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 8F
10:48:19.515 -> DATW 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD E 0
10:48:19.515 -> DATW 48 A2 65 A2 6C A2 6C A2 6F A2 20 A2 57 A2 6F A2 72 A2
6C A2 64 A2 21 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 2
10:48:19.515 -> DATW 43 A2 FF A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 1D 8 89
10:48:19.515 -> DATR 0 A2 0 4
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD 11
10:48:19.515 -> DATR 0 A2 0 1
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.515 -> CMD D 8 89
10:48:19.515 -> DATW 4 A2
10:48:19.515 ->
10:48:19.515 -> CMD 83
10:48:19.515 -> DATW 0 A2 0 A2 0 A2
10:48:19.515 ->
10:48:19.515 -> CMD C0
10:48:19.515 -> DATR 0 22 0 22
10:48:19.515 ->
10:48:19.752 -> CMD 2
10:48:19.752 -> DATW 43 A2 FF A2
10:48:19.752 ->
10:48:19.752 -> CMD C0
10:48:19.752 -> DATR 0 22 0 22
10:48:19.752 ->
10:48:19.752 -> CMD 80
10:48:19.752 -> DATW 0 A2
10:48:19.752 ->
10:48:19.752 -> CMD C0
10:48:19.752 -> DATR 0 22 0 22
10:48:19.752 ->
10:48:19.752 -> timeout!

**Comment by lvplummer** [comment]
I mentioned earlier that my transmitter is very simple so that I can
minimize current draw during sleep - UNO DIP chip, crystal, 2 caps, SX1262
module, and 2 AAA batteries.  I tried different UNO chips to make sure the
issue was not the CPU.  That caused me to focus on the radio module.

By doing some timing measurements using the CPU millis() timer versus
external oscilloscope, I determined that the CPU was running at about 4 MHz
instead of 16 MHz.

I replaced the bad crystal and all boards with radios from two sources
running on two bands are working as expected.
